File: oscar-algo/src/main/java/oscar/algo/reversible/Reversible.java
/*******************************************************************************
 * OscaR is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 2.1 of the License, or
 * (at your option) any later version.
 *   
 * OscaR is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License  for more details.
 *   
 * You should have received a copy of the GNU Lesser General Public License along with OscaR.
 * If not, see http://www.gnu.org/licenses/lgpl-3.0.en.html
 ******************************************************************************/
package oscar.algo.reversible;


/**
 * Generic Reversible inside a reversible node
 * @author Pierre Schaus pschaus@gmail.com
 */
public abstract class Reversible {
	
	private int lastMagic;
	public final ReversibleContext node;
	
	public Reversible(ReversibleContext node) {
		this.node = node;
		lastMagic = -1;
	}
	
	protected boolean mustBeTrailed() {
		if (lastMagic == node.magic()) {
			return false;
		} else{
			return true;
		}
	}
	
	protected void trail() {
		if (!mustBeTrailed()) return;
		else {
			lastMagic = node.magic();
			addOnTrail();
		}
	}

    /**
     * Add what encapsulates the state of the object on the trail
     */
	protected abstract void addOnTrail();

    /**
     * Restore the state of the object
     * @param val , an object representing the state necessary for a restoration
     */
	protected abstract void restore(Object val);

}

File: oscar-algo/src/main/scala/oscar/algo/reversible/Reversible.scala
package oscar.algo.reversible

/**
 * Generic Reversible inside a reversible node
 * @author Pierre Schaus  pschaus@gmail.com
 * @author Renaud Hartert ren.hartert@gmail.com
 */
abstract class Reversible[T](val context: ReversibleContext) {
  
  private var lastMagic: Long = -1L

  @inline protected final def trail(): Unit = {
    val contextMagic = context.magic
    if (lastMagic != contextMagic) {   
      lastMagic = contextMagic
      context.pushOnTrail(this, value)
    }
  }
  
  /** Return the current value of the reversible */
  def value: T
  
  /** Restores the state of the object */
  def restore(value: T): Unit  
}